### PR TITLE
Smt checker

### PR DESCRIPTION
**Pull request by Alex-Fischman**
Adds an SMT-based bounded model checker. This is useful because it doesn't require specifying universe bounds, however it's very slow.

### CODE REVIEW

File: bounded/src/smt.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
+// Copyright 2022-2023 VMware, Inc.
+// SPDX-License-Identifier: BSD-2-Clause
+
+//! A bounded model checker for flyvy programs using the [CaDiCaL][cadical] SAT solver.
+//!
+//! [cadical]: https://fmv.jku.at/cadical/
```
Comment (tchajed) [inline bounded/src/smt.rs:6]: This doc comment seems wrong, this is an SMT checker and not SAT.

File: bounded/src/smt.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        (forall N:node. unlock_msg(n) &
+            (((unlock_msg(N))') <-> unlock_msg(N)) &
+            ((server_holds_lock)')) &
+        (forall x0:node. ((lock_msg(x0))') = lock_msg(x0)) &
+        (forall x0:node. ((grant_msg(x0))') = grant_msg(x0)) &
+        (forall x0:node. ((holds_lock(x0))') = holds_lock(x0)))
+
+# safety:
+assert always (forall N1:node, N2:node. holds_lock(N1) & holds_lock(N2) -> N1 = N2)
+        ";
```
Comment (tchajed) [inline bounded/src/smt.rs:260]: Can you move the flyvy input files to their own files and load them here from disk? It's okay if they aren't also integration tests.

**Review by tchajed — COMMENTED** [review]
I have some minor comments, otherwise looks good.

Could you also update the PR description to describe what you benchmarked and the results you got?

File: bounded/src/smt.rs (hunk 2 above)
Comment (Alex-Fischman) [reply bounded/src/smt.rs:260]: I'll do this in it's own PR, all the checkers share tests.

**Comment by Alex-Fischman** [comment]
I didn't do any benchmarking; I just know it's slow because the lockserver_buggy test takes longer than any of the others (I `#[ignore]`d it.)

**Comment by tchajed** [comment]
Please rebase onto `main`, then I can merge.

**Comment by tchajed** [comment]
Sorry, this needs another rebase due to the other PR being merged :)

**Pull request merged**